app/src/main/java/me/quadphase/qpdex/DetailedPokemonActivity.java

```diff
@@ -451,21 +454,94 @@ public void onClick(View view) {
      * Fills in how the different types affect the detailedPokemon.
      */
     private void fillTypeComparisonInfo(){
-        //TODO: Replace with real logic when the sparse type matrix is available!
-        List<Type> typeList = new ArrayList<Type>(3);
-        typeList.add(new Type("Normal", 13));
-        typeList.add(new Type("Electric", 4));
-        typeList.add(new Type("Ice", 12));
-
-        LinearLayout testy = createTypeMatchBlock("Immune",typeList);
-        testy.setBackgroundResource(R.color.dex_blue_transparent);
-        LinearLayout testy2 = createTypeMatchBlock("Resists",typeList);
-        testy2.setBackgroundResource(R.color.dex_blue_transparent);
-        LinearLayout testy3 = createTypeMatchBlock("Resists",typeList);
-        testy3.setBackgroundResource(R.color.dex_pink_transparent);
-        typeStrong.addView(testy);
-        typeStrong.addView(testy2);
-        typeWeak.addView(testy3);
+        // determine the effectiveness of each type against the pokemon in question. To do this
+        // on a pokemon with just one type, use the information directly. If the pokemon has 2
+        // types, multiply the attacking effectiveness of both types to determine the final values.
+
+        // This pokemon is immune to these types when defending itself (i.e. 0 effectiveness)
+        List<Type> typeImmune = new LinkedList<>();
+        // This pokemon is resistant to these types when defending itself (i.e. 1/2 effectiveness)
+        List<Type> typeResist = new LinkedList<>();
+        // This pokemon is very resistant to these types when defending itself (i.e. 1/4 effectiveness)
+        List<Type> typeResistStrong = new LinkedList<>();
+        // This pokemon is weak to these types when defending itself (i.e. 2 effectiveness)
+        List<Type> typeWeakTo = new LinkedList<>();
+        // This pokemon is very weak to these types when defending itself (i.e. 4 effectiveness)
+        List<Type> typeWeakStrong = new LinkedList<>();
+
+        // get the types of the pokemon
+        List<Type> types = detailedPokemon.getTypes();
+
+        // get the list of all the types
+        Type[] allTypes = Type.getListOfTypes();
+
+        // when the pokemon has one type:
+        if (types.size() == 1) {
+            Type type = types.get(0);
+            for (int i = 1; i < Type.getNumberOfTypes() + 1; i++) {
+                double effectiveDefence = type.getDefendingEffectivenessAgainst(allTypes[i]);
+                double effectiveOffence = type.getAttackingEffectivenessAgainst(allTypes[i]);
+
+                // add the types to the correct lists
+                if (effectiveDefence == 0)
```

Comment:
I have. Unfortunately, switch does not work on doubles, and we have values of 0.25 and 0.5.